scripts/generate-icons.ts

```diff
@@ -0,0 +1,285 @@
+import fs from 'fs/promises';
+import { createCanvas, loadImage } from 'canvas';
+import { resolve } from 'path';
+import { getBrowser, releaseTarget } from './util';
+import colorLog from './log';
+import { PluginOption } from 'vite';
+
+/**
+ * Input folder containing the original svg icons in the source.
+ */
+const input = resolve('', 'src', 'icons');
+
+/**
+ * Output folder where the rendered images should be output during the building.
+ */
+const output = resolve('', 'build', getBrowser(), 'icons');
+
+/**
+ * Resolutions to render main icon to.
+ */
+const mainIconResolutions = [48, 96, 128, 256, 512];
+
+/**
+ * Resolutions to render action icons to
+ */
+const actionIconResolutions = [16, 19, 32, 38];
+
+/**
+ * A map of filenames in the misc icon folder and the sizes we want them in PNG format
+ */
+const miscSizes = {
+	'icon_generic.svg': [16],
+};
+
+/**
+ * A map of filenames in the monochrome icon folder and the colors we want them in PNG format
+ */
+const monochromeColors = {
+	'action_base.svg': {
+		light: '#d51107',
+		dark: '#fc3434',
+	},
+	'action_disabled.svg': {
+		light: '#7f7f7f',
+		dark: '#9f9f9f',
+	},
+	'action_error.svg': {
+		light: '#fea800',
+		dark: '#fea800',
+	},
+	'action_ignored.svg': {
+		light: '#fea800',
+		dark: '#fea800',
+	},
+	'action_loading.svg': {
```

Comment:
For me this one I couldn't see - it was the same colour as my browser .

![image](https://user-images.githubusercontent.com/713525/230981249-d813a19b-2dcb-4fda-9546-3427e6770eef.png)

Reply:
Interesting, that should be white since you have dark theme enabled. Could you try to repro and look for errors in the service worker?

Reply:
I updated the theme fetching and it should be more reliable now I think.

Reply:
Not sure what is up :/ Something with the transition with the action item causes the icon to go away for me until it's resolved :/ very odd.

Reply:
Okays - it's nothing to do with this.. It's some performance issue on my end - if I change `getIconType` to just return `dark` it performs as expected. something with `getBrowserPreferredTheme` is causing it to lag loading :/

Reply:
i did - the loading icon never showed unless i made that function static value return. I'll record a video tonight.

Reply:
Can you run the following functions in any tab (not service worker) and show results
`window.matchMedia('(prefers-color-scheme: dark)').matches`
`window.matchMedia('(prefers-color-scheme: light)').matches`

and then run the following function in service worker or extension settings/popup and show results:
`chrome.storage.local.get("StateManagement", console.log)`

Reply:
```
window.matchMedia('(prefers-color-scheme: light)').matches
true
window.matchMedia('(prefers-color-scheme: dark)').matches
false
```

So I'm guessing I'm running in light mode by default. 

```
chrome.storage.local.get("StateManagement", console.log)
undefined
```


Here's a video of the behaviour on latest branch https://user-images.githubusercontent.com/713525/231571816-b82845cc-2220-4b71-b2c4-6b3d0282f269.mp4

Reply:
Looked into it, turns out prefers-color-scheme just [straight up doesnt work in chrome since 3 years ago](https://bugs.chromium.org/p/chromium/issues/detail?id=998903) ;-;
Apparently a fix is being pushed to version 114, _maybe_ 113.

That sucks. We can't get the users preferred theme for a while then.

Reply:
I guess i change the colors up so that we can afford dealing with chrome being terrible at following web standards, task for tomorrow though

Reply:
Strange - I often wondered why my UI was dark yet was always served a light theme :D Probably some linux quirk too as I suspect it runs fine on other OS

Reply:
Yeah based on the issue it seems to be a linux issue

Reply:
Could be a tough one as Chrome supports the user customising the browser colour. 

I just changed it to something less dark and it works perfectly.

Reply:
Think I have an idea to failsafe this, gonna be a slight pain to implement though but shouldnt be too bad

Reply:
Alright so this now creates borders on the icons. It's not exactly pretty if your browser theme color is far off the expected color for your preferred theme, but it does completely eliminate any possibility of issues occurring, and is hardly visible with standard settings (other than linux chrome with dark theme until chrome 114). Safari icons unaffected by this change.

Reply:
Seems legit to me from testing :)